### ISSUE TITLE
I am lazy and don't feel like uploading images to github :')

### ISSUE THREAD

**Issue by Velythyl** [Velythyl/duckietown-internship-progress#1]
# Questions from Duckietown

![image](https://user-images.githubusercontent.com/31974070/83983171-2b4d5480-a8fa-11ea-8956-be82df7c2ed6.png)

## Human expert:  
+We can be certain that the samples are as good as they can be (in simple problems).  
+We can tweak the expert behaviour to make sure the policy can learn well. For example, pure pursuit turns "late", so it's hard for the policy to see identifying features of turns when the pictures for left and right turns are the same - because of how late the turn is.  
-It's time consuming for the humans.  
-Because it is so time demanding, it boils down to training with logs, as you can't ask a human to offer samples any time you require them.  

## Logs:  
+Since they're static logs, it's possible to parrallelise the sampling step. One could simply ask a hundred contributors, or a cluster to run a bunch of simulations, and then gather all that up to have a huge amount of data.   
+Allows for mixing our different strategies: whether it's a human, ground-truth, or learned expert, they can all produce logs, and those logs can all be mixed together to reduce overfitting.  
-Should the API change, all those hard-collected logs become unusable.  
-It's difficult to make sure the logs are good. You would have to restart an env with the same seeds, and then run the actions from the logs, to make sure the agents are not being fed bad actions.  

## Learned policy:  
The cons outweigh the pros by a lot here, in my opinion  
-If we're learning from a learner, why not simply duplicate the learner and use that instead?   
-The ground truth experts are better than learned policies in almost all cases, so training on something that learned from the ground truth experts would introduce the same errors learned by the learners already. In other words, it's kind of the same reason why in image classification you don't add your predictions to your training dataset: training on results of training introduces error. Of course, with dagger, we can mitigate that by using the student more and more and making it learn how to recover, but why not start from a "flawless" ground truth expert instead?  
+Perhaps this allows us to stabilise RL policies: we start from a flawed RL policy that actually learned a behaviour that our ground truth experts or humans can't do well, and then we learn that quirk with IL, and then we learn the rest of the desired behaviour with other experts. That way, we get the benefits of the problem space exploration methods of RL with less drawbacks.  
+If by "training on a learned expert" we mean something akin to transfer learning, then those cons disappear. We could use the image segmentation and feature recognition of a very well trained and optimised model, and simply add a layer over it that translates a vector of features into our desired actions.  

## Hard-coded expert:  
+Since in most cases this expert has access to the ground truth, it's very, very performant, rivalling humans. This means that we have access to a human-like expert, without the drawback of having a human strapped to a chair for days on end generating experiment logs for our IL training.  
+The expert *can* be tweaked to offer exactly the type of behaviour we want from the IL model  
-The expert *has to* be tweaked to offer exactly the type of behaviour we want from the IL model, and this can be time consuming, and involve complicated code just to make it work.  
-The expert acting on the ground truth can be a boon in some cases, but a hindrance in others. If the expert is "too good" at using the ground truth, it can be impossible for the IL training to duplicate that behaviour, since the information for triggering it can be completely hidden from IL. This means that since we're penalising the IL for not acting as the expert does, the IL model could learn to do the action for the wrong reasons, just because the expert *tends* to do the tricky action when an unrelated image feature is present.  

# Dagger/IL questions  

## Bailing  

I feel like IL is very prone to bailing when it encounters something that the expert has never encountered before.  
What is the best method to fix this? We can run residual learning on the IL model once it reaches good performance in expert-adequate tasks, and purposefully put it in bad situations so that the residual learning can fix the bad behaviour (which is a thing we discussed before).  
But I had the idea of using logs in addition to using the expert on the go, and of introducing logs of a "extreme situations" expert, who finishes his run as soon as he has finished getting back on track. Like, spawn the duckiebot in a very bad spawn point, and then use the expert to get it back on track, or have human logs of the same situation.

I feel like both could work, and both have pros and cons:

### Residual:  
+Less supervised. Can just infer actions based on the reward function  
+Proved to work, since the student's RPL repo was doing just that and working very well
-Less stable; might not always converge  
-Much slower  

### "Dangerzone" logs:  
+We know dagger is super good at learning from expert behaviour, so it stands to reason it'd be good at learning how to get back on track too  
+Can be included directly in the current training loop, while residual learning must be applied after IL has trained  
-Might cause stability issues if the dangerzone images are too similar to on-track images  
-It's just a random idea I had, so it's untested  

## Bigger history  
Sometimes, the bot fails to turn, slows down until it's up against the wall, and then just bails. But I feel like it had no choice to do so: even I, a real human person thing, could not guess which direction I should turn based on those last images. They're indistinguishable from left or right turns, or even just starting in a dangerous position perpendicular to a lane.  

Reading the paper https://spectrum.library.concordia.ca/984373/1/Diaz_MSc_F2018.pdf , I came across this passage:

![image](https://user-images.githubusercontent.com/31974070/84092196-a67b3d00-a9c4-11ea-9d82-8df4bbd0d57c.png)

And from what I've seen from the Deepmind RL videos, "Hence, a set of
realistic assumptions are required to reduce the complexity of the analysis of any MDP" tends to actually mean "we ignore the history, and simply look a ONE observation".

I feel like that works well in most cases, because those few cases I described above only occur if the learner doesn't exactly match the expert. But, well, that doesn't always happen. And perhaps keeping a small history would help us.

I don't want to keep the entire history, and I don't think keeping the last few frames would help either, since those are usually just as featureless as the latest one.

I'd like to keep a buffer of maybe 5 frames, spaced with x frames intervals. That way, if we do get to one of the dangerous positions, I'm guessing that having a frame that clearly shows a curve to the right would help us get back on track. 

I'd probably put really higher weights on the latest frame, so that the others aren't used a lot. Or maybe even make a "danger zone detector" which summons a second learner whose sole job is parsing the history to get out of the dangerous zone.

The specifics of the architecture will come later, but does this sound like a good idea? It wouldn't be a RNN per say, because we don't care about previous actions, we're just hoping to work on 5x3x160x120 images instead of 1x3x160x120... with the first of those dimensions being the past.

**Comment by Velythyl** [comment]
Re: using an expert to parse dangerous situations

I've just started reading the paper, and it seems to aim to solve that directly! I'll note here if my idea is still applicable once I'm done reading.

**Comment by bhairavmehta95** [comment]
> I'd like to keep a buffer of maybe 5 frames, spaced with x frames intervals. That way, if we do get to one of the dangerous positions, I'm guessing that having a frame that clearly shows a curve to the right would help us get back on track.

It's pretty cool that you came to this conclusion yourself! This is actually a very standard practice, called [frame stacking](https://github.com/openai/gym/blob/master/gym/wrappers/frame_stack.py) - it's how we encode history (we even do this with state-based observations - i.e when it is a vector rather than image). It's also much easier than learning an RNN / LSTM.

**Comment by bhairavmehta95** [comment]
>  Like, spawn the duckiebot in a very bad spawn point, and then use the expert to get it back on track, or have human logs of the same situation.

This is also a good idea, what is one issue you might run into here? (Hint: what's a programmatic way to define bad)? Is there any way you could potentially generate these 'checkpoints' easier (Hint: you have the entire trajectory, and you also know when the environment is `done`).

**Comment by bhairavmehta95** [comment]
> Since in most cases this expert has access to the ground truth, it's very, very performant, rivalling humans. This means that we have access to a human-like expert, without the drawback of having a human strapped to a chair for days on end generating experiment logs for our IL training.

Is this true? Or are we often working on too simple of problems

**Comment by Velythyl** [comment]
> > I'd like to keep a buffer of maybe 5 frames, spaced with x frames intervals. That way, if we do get to one of the dangerous positions, I'm guessing that having a frame that clearly shows a curve to the right would help us get back on track.
> 
> It's pretty cool that you came to this conclusion yourself! This is actually a very standard practice, called [frame stacking](https://github.com/openai/gym/blob/master/gym/wrappers/frame_stack.py) - it's how we encode history (we even do this with state-based observations - i.e when it is a vector rather than image). It's also much easier than learning an RNN / LSTM.

Oh, very nice! I'll read up on that tomorrow, it would be cool to include it, as I think it could help 

> This is also a good idea, what is one issue you might run into here? (Hint: what's a programmatic way to define bad)? Is there any way you could potentially generate these 'checkpoints' easier (Hint: you have the entire trajectory, and you also know when the environment is done).

I actually noticed the problem with unparseable curves because of how the environment reacts when I'm testing the trained policies - when the bot fails, it spawns it very close to the point where it died, but often in very precarious positions. So I planned on using the env's "you just died; please retry" logic to generate bad spawn points.

However, your way makes much more sense. When working on the RL repo, I tried coding a (very bad) Stanford controller, so I know the env can give me the current angle + position. I could do something like

1. Generate normal spawn coordinates and angle
2. Project a point like the pure pursuit controller does
3. Use the current angle and the trajectory angle to turn the bot in the wrong direction

I could also make it not die on `done`, and reset the env to a few steps prior, and then remove any control from the leaner for a few steps. That way is a lot less intrusive/requires less code, but it relies on chance.

> Is this true? Or are we often working on too simple of problems

That's fair! I should have thought of that

I guess it also depends on how studied the problem has been before. For some, we have no expert programs at all, even with ground  truth.

**Comment by Velythyl** [comment]
# Residual Learning (RPL) questions

https://arxiv.org/abs/2004.00993  

First off, I read this passage in the RL paper you linked me, and I liked that they also talked about the "magic threshold". 

![image](https://user-images.githubusercontent.com/31974070/84823419-4c91fd00-afec-11ea-95b3-5ab59d6a2faa.png)

+ But said paper didn't talk about checkpointing, even though you mentioned it. It seemed like a pretty important part of RPL. Is there another paper that explains how that works?  

https://arxiv.org/pdf/1905.11108.pdf

+ That paper talked about initialising the replay buffer with the samples collected from the expert before doing the IL. Could we do the same thing, basically by not resetting the replay buffer, and using it directly from RL to RPL?

https://arxiv.org/pdf/2006.04678.pdf  

+ Interestng, but not linked to RPL. I think if we choose to try the "bad faith actor" idea, it might be interesting to compare to it.

**Comment by bhairavmehta95** [comment]
Sorry, how does expert regret = magic threshold?

**Comment by Velythyl** [comment]
Well, the way they define it is not "magic", but it felt to me like what we had talked about. 

I didn't have the theoretical foundations to describe it, but I think they describe the same behaviour of "past a certain loss, the learner does well, but the loss stops to have a meaningful impact on performance".

Like we had said: get to that threshold, stop IL, and then start RPL
